### ISSUE TITLE
Support Terraform data sources

### ISSUE THREAD

**Issue by sergiught** [auth0/terraform-provider-auth0#22]
From terraform-provider-auth0 created by [lebaptiste](https://github.com/lebaptiste): alexkappa/terraform-provider-auth0#279  

### Description

Aside Terraform resources, there is also another [data source](https://www.terraform.io/docs/configuration/data-sources.html) type which is roughly the readonly version of a resource.  

That would be a great enhancement to this provider to include data source support. Since I have already worked on a terraform provider I am happy to submit a PR for it.
 
### New or Affected Resource(s)

Ideally all the existing resources would have their data source counterpart.

### Related Issues

- https://github.com/auth0/terraform-provider-auth0/issues/60
- https://github.com/auth0/terraform-provider-auth0/issues/414

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by paullalonde** [comment]
Here's some rationale behind supporting data sources: they are very important when you have Auth0 resources provisioned across multiple Terraform configurations. For example, one Auth0 tenant could have a Connection provisioned in one Terraform configuration, and multiple Auth0 Clients provisioned each with their own Terraform configuration. The data source in this case can be used to supply information to the TF code that creates the Auth0 Client.

Personally, it would help me a lot if there were data sources for Connection and Resource Servers. I'm willing to contribute them.

**Comment by Nmishin** [comment]
It will be really useful if we can grab [this ip addresses](https://auth0.com/docs/secure/security-guidance/data-security/allowlist#united-states) via data resource.
Because for now we updated our firewall by hands, from time to time ...

**Comment by arogic** [comment]
Hi, is there any progress on the User data source request? Regards

**Comment by sergiught** [comment]
Hey folks 👋🏻 

A quick update from our side, we have this scheduled to work on in the next month and it should be delivered by the end of Q1 if not earlier, if changes arise we'll updated this issue. 

It is very unlikely we'll be able to add a data source equivalent for ALL resoucers. At the moment the following are considered:

- user
- tenant
- resource_server
- connection
- organization

Would you see the need for any other data source that would make your lives easier? 🙏🏻 Feedback is encouraged and appreciated!

**Comment by sergiught** [comment]
Hey folks, we're slowly adding all the requested data sources. If anyone wants to test some of them before we cut a new release you can do so by running `make install VERSION=0.43.1` after cloning this repo. You only need Go (1.18+) and Terraform (1.0+) installed to do so. 

Then reference in your terraform main file the following:

```
terraform {
  required_providers {
    auth0 = {
      source  = "auth0/auth0"
      version = "0.43.1" # The same version we installed locally
    }
  }
}

provider "auth0" {
  debug = true
}
```

Rough ETA for a release ~ end of February / early March.

**Comment by sergiught** [comment]
Hey folks, we're gonna cut the release with support for data sources towards the end of this week. 

Data sources added:
- connection
- organization
- resource server
- tenant
- user
- role
- attack protection
- branding
- branding theme

**Comment by sergiught** [comment]
Hey folks, a bit ahead of schedule we just released [v0.44.0](https://registry.terraform.io/providers/auth0/auth0/latest/docs) with support for the above mentioned data sources. Feedback is encouraged and appreciated! If you encounter any issues with them please let us know.

**Comment by sergiught** [comment]
Hey @marcportabellaclotet-mt 👋🏻 

Adding a `users` attribute on the `auth0_role` data source is not currently on our roadmap, would you mind opening a separate GitHub issue for this? 🙏🏻 So we can at least have it on our radar for the future. 

Realistically I can't give you an ETA for when we'll be able to achieve this as we need to be really mindful of performance issues considering an Auth0 Tenant could have hundreds and hundreds of thousands of users.

At the moment we're working on pushing this provider to a stable v1 that should roll out in ~1-2 months.
